### DISCUSSION TITLE
RFC: Versioning

### DISCUSSION THREAD

**Discussion by JsantosDK** [help-me-mom/ng-mocks#9730] [General]
### Background
---
When ngMocks was created, it was just a normal testing library, as time moved on, it improved into being one of the most used libraries for testing in the angular ecosystem. However, its development doesn't follow the normal development from the majority of Angular libraries like Angular Material and others.

As you can see in the official docs, the latest version can be tested against an Angular 5 application.

| Angular | ng-mocks | Jasmine | Jest | Ivy |
|--------:|:--------:| :-----: | :--: | :-: |
|      18 |  latest  |   yes   | yes  | yes |
|      17 |  latest  |   yes   | yes  | yes |
|      16 |  latest  |   yes   | yes  | yes |
|      15 |  latest  |   yes   | yes  | yes |
|      14 |  latest  |   yes   | yes  | yes |
|      13 |  latest  |   yes   | yes  | yes |
|      12 |  latest  |   yes   | yes  | yes |
|      11 |  latest  |   yes   | yes  | yes |
|      10 |  latest  |   yes   | yes  | yes |
|       9 |  latest  |   yes   | yes  | yes |
|       8 |  latest  |   yes   | yes  |     |
|       7 |  latest  |   yes   | yes  |     |
|       6 |  latest  |   yes   | yes  |     |
|       5 |  latest  |   yes   | yes  |     |

---
## Proposal
I would like to propose that ngMocks follows the Angular team versioning logic such that we would have:

| Angular | ng-mocks | Jasmine | Jest |
|--------:|:--------:| :-----: | :--: |
|      20 |  20  |   yes   | yes  |
|      19 |  19  |   yes   | yes  |
|      18 |  14  |   yes   | yes  |

-- 
## Advantages

1. Following the standard versioning system
The current versioning system doesn't exist for this library, it tries to keep all possible versions happy. This increases the difficulty to add new features or fix changes that occur when the Angular Team adds big functionalities, example: the addition of the new `input.required` that created several issues in the repository.

2. Less need for distinct behavior between versions
With a focus on the last 3 versions, there is less of a need to test against old specs for errors that break only in old versions of Angular that are not unsupported. This does not mean that the library would not add patches to these versions (see next advantage).

3. Better bug prioritization
A bug on the latest version of Angular is not as important as a bug on an older and unsupported version.
This would help maintain the library since bugs in old version could take a reduced importance versus new changes that will remain going forward.

4. Reduced dimension and easier maintainability
This library tests its functionality versus several issues, some of which only occur in very old versions of Angular that are no longer supported. Keeping these files for newer versions doesn't bring much value and adds dimension to the library. 
With the goal of making the library easier to maintain, reducing the dimension would help new developers to contribute and lighten the weight on @satanTime.

5. More predictable package.json behavior
It is a bit strange well Angular changes from version x.y.z to x+1.0.0, and the library keeps the same version.
For users who are migrating their applications, some may lose time checking if the library keeps up to date. By following the same version as the Angular team for the main branch, it makes the migration behavior more predictable for existing applications.

**Comment by satanTime** [comment]
Hi. 

There is no need to follow angular versions, because ng-mocks doesn't depend on angular implementation and works with any angular version (apart from latest features which should be added to ng-mocks).

Semantics is used in case of dependency when each lib version depends on a specific angular implementation, whereas that's not the case for ng-mocks.

**Reply by JsantosDK** [reply]
I really don't agree on this.
You do version this library according to angular, hence going from `14.12.2` to `14.3.0`.
In `14.12.2`, you have code that is specific to Angular 17, as you point out in the commit.
And this is often seen in the CHANGELOG.md file.

For all intends and purposes, the lib is versioned according to Angular releases.
This makes sense since on nearly all every Angular release, there are breaking changes.
https://github.com/ngneat/spectator, which is probably the other most used testing library follows this versioning logic as well for some of the motives I mentioned.

**Reply by satanTime** [reply]
That's not very accurate. You can check the number of weekly downloads. 

If angular 18 / 19 will be breaking changing for ng-mocks, then it will be a major release. 

Think about that in a different way than just semantic of angular. 

Someone is on angular 14 and uses latest ng-mocks. Angular 19 is out, no braking changes for ng-mocks. However, if it's going to be a major release, it means all new features should be after that major release too, because to maintain alone all different release branches is much more complicated and time consuming than just one. 
So, that dev simply hesitates to update ng-mocks to the next major to get new testing features and optimizations locking him/her- self in legacy for no reason.

With the current approach, they simply update ng-mocks to the next minor release and they get all new features.

**Reply by JsantosDK** [reply]
I understand this logic, but on every angular releases there is a need to make some fixes on the ngmocks lib. 

Furthermore, as a consuming dev, if I am on angular 14, I don't care about the latest. I care about the version that is compatible with angular 14. More so, if I'm on a given version, I don't care about changes in other versions, above or below. The only time I care is when I need to migrate to a new major, and at that point, I'm expecting to migrate other dependencies as well. So I don't see the point you are trying to make.

This also doesn't help the fact that if angular adds something in version 19 that breaks version 14, that the lib needs to manage both cases, when the number of users on version 14 is very small.

The only point that I can understand where it is beneficial is in that you only need to maintain one branch. But even that I wouldn't really agree. That to me is a sign that another problem exists (either the lib structure is too complex, or more help is needed to maintain).

**Reply by satanTime** [reply]
Let's put it in a different way.

What is the benefit of the next approach?
1. feature 1 has been added to ng-mocks for all angular versions
1. Angular 19 release, breaking change of ng-mocks without real breaking changes in the lib itself
1. feature 2 has been added to ng-mocks for all angular versions

Why step 2 should cause major release and block anyone on prev version from using the feature 2 on their current angular?

**Reply by satanTime** [reply]
I understand semantic releases in `ngrx` which depends on its angular core. In case of `ng-mocks`, there is no dependency on angular.

**Reply by JsantosDK** [reply]
After the second step, there are often fixes that need to occur that are related to Angular 19 specifically, not related to other versions.

There is a difference between fixing something and adding a new functionality, and I think this is where we diverge.
You are focusing on new features being added, I'm focusing on fixing stuff.

At this point, one can look at ngMocks, and not a lot of features are being added, if any.
Most PR's that are actual PR's are fixes concerning new angular versions.

The flow you mentioned makes sense from a feature perspective. Having versioning matching the Angular version does not bring a benefit here. I won't disagree on that.

But now, I will go with the opposite case, tell me what is more logical:

1. Issue Foo occurs in Angular 16 only;
2. Angular 19 is released, thus Angular 16 is no longer maintained;
3. Issue Foo solution is pushed into all versions of Angular even if it doesn't make sense for older versions;

vs 

1. Issue Foo occurs in Angular 16 only;
2. Angular 19 is released, thus Angular 16 is no longer maintained;
3. Issue Foo solution is pushed in branch for version 16, since it only occurs there, other branches are not impacted;

The first option will force you to keep adding checks that will ultimately make it a Angular version check in the code.
The second version means that you can reduce the size of the library as Angular removes/deprecates functionality and allows the code to be smaller/cleaner.

At the end of the day, this is why I think we are talking about different approaches.
To me, ngMocks will not release new features often, instead it will accompany changes in the angular ecosystem.
Matching changes like like defered views and input.required changes, or just flushing of effects are not new features of ngMocks, they are new requirements to match Angular version's.

**Reply by JsantosDK** [reply]
Also, I never mentioned `ngrx` and I don't think it is a good case to compare, for one, it is not even a testing library.
If you want to use a case to support your goal, I would say https://github.com/ngx-translate/core makes much more sense.
But again, that is not a testing library so it doesn't make sense for me to compare with that in either case.

**Reply by satanTime** [reply]
I wish `Angular 16 is no longer maintained` could be true.

There are enterprises which are on very old angular versions and don't follow the way Google deprecates Angular. In the same time, I don't get why they should jump from `ng-mocks@12`to `ng-mocks@19` if they need a new feature from there, whereas in `ng-mocks` itself hasn't been any breaking changes.

**Reply by JsantosDK** [reply]
Unfortunately, I know that there are companies that don't follow the normal migration timeline, that is not my point. 
But that is not the point. 

I'm going use the reverse example. 
Imagine that it is decided to introduce a feature related to signals in ng-mocks, like a wrapper to TestBed.flushEffects.

This functionality would not make sense in a version without signals. However, it would be exposed by ng-mocks for those applications. This could generate confusion or other issues.

**Reply by satanTime** [reply]
This is the way `ng-mocks` releases new versions: https://semver.org/ 

> Following the standard versioning system

in case of `ng-mocks`, to support a new version of angular and don't break the support of prev version doesn't require much more time.

> Less need for distinct behavior between versions

it doesn't work for enterprises, there are many companies which still use very old angular versions. For `ng-mocks`, it doesn't cost much to support old angular versions.

>Better bug prioritization

In a company - yes. In my case, it's more about availability to fix something in spare time. Unfortunately, as you can see, I've been busy for quite a while. And, I hope it is going to be changed soon, so I can contribute more. Until then an easy fix will be of a higher priority due to the less time required.

> Reduced dimension and easier maintainability

Again - it doesn't cost much to support older versions.

> More predictable package.json behavior

`ng-mocks` doesn't depend on Angular, it simply creates mock declarations. Consider `ng-mocks` as `zone.js` or `rxjs`.

**Reply by JsantosDK** [reply]
So basically the issue is that what I'm requesting requires to much work for you. Which is funny soccer to me, I would like to contribute to this lib, but having to install dependencies every day (constant pr`s for minor stuff) and just running this lib, is a nightmare. 

There is no explanation in the code for why something's are how they are, there is no model that explains the idea behind stuff. 

It has a very high barrier of every for a new contributor. As a result, you get the situation you are in now. You are the only active maintainer and I, who want to contribute, don't because I don't understand several decisions of the lib and everything is dependent on you. 

As an example, I wanted to contribute with schematics, but never got a response.

My goal with this discussion was to be able to remove stuff from the lib, so that others like myself can contribute to it. 

I will stop annoying you since there doesn't seem to a way to convince you and you're already with a lack of time.
